frail/sharing_nested_counters_rewrite.py

```diff
@@ -0,0 +1,164 @@
+    if isinstance(lake_state.program_map[index], CounterSeq):
+        prod_counter = lake_state.program_map[index].producing_counter
+        can_merge = True
+        for m in other:
+            if e.index in other[m]:
+                can_merge = False
+        if can_merge:
+            if prod_counter in curr:
+                curr[prod_counter].append(e.index)
+            else:
+                curr[prod_counter] = [e.index]
+    sharing_nested_counters_rewrite(
+        lake_state.program_map[index], False, lake_state)
+
+
+def sharing_nested_counters_rewrite(
+        e: AST,
+        root: bool = True,
+        lake_state: LakeDSLState = default_lake_state):
+    global cur_scan_idx, output_scan_index, cur_scan_lambda_var
+    global printed_ops, have_merged_counters, prev_vals, max_vals
+    global replace
+
+    if root:
+        cur_scan_idx = -1
+        output_scan_index = -1
+        printed_ops = set()
+        have_merged_counters = False
+        prev_vals = {}
+        max_vals = {}
+        replace = {}
+
+    e_type = type(e)
+    if e.index in printed_ops:
+        return
+
+    printed_ops.add(e.index)
+
+    if e_type == RecurrenceSeq:
+        old_scan_idx = cur_scan_idx
+        old_scan_lambda_var = cur_scan_lambda_var
+        sharing_nested_counters_rewrite(
+            lake_state.program_map[e.producing_recurrence], False, lake_state)
+        cur_scan_idx = old_scan_idx
+        cur_scan_lambda_var = old_scan_lambda_var
+
+    elif e_type == CounterSeq:
+        counter_op = lake_state.program_map[e.producing_counter]
+        old_scan_idx = cur_scan_idx
+        old_scan_lambda_var = cur_scan_lambda_var
+        sharing_nested_counters_rewrite(counter_op, False, lake_state)
+        cur_scan_idx = old_scan_idx
+        cur_scan_lambda_var = old_scan_lambda_var
+
+    elif e_type == CounterOp:
+        if output_scan_index == -1:
+            output_scan_index = e.index
+        cur_scan_idx = e.index
+
+        if e.prev_level_input is not None:
+            add_potential_counter(e, lake_state, e.prev_level_input, True)
+
+        if e.is_max_wire:
+            add_potential_counter(e, lake_state, e.max_val, False)
+
+        if isinstance(e.incr_amount, AST):
+            sharing_nested_counters_rewrite(e.incr_amount, False, lake_state)
+
+    elif e_type == ScanConstOp:
+        if output_scan_index == -1:
+            output_scan_index = e.index
+        cur_scan_idx = e.index
+        cur_scan_lambda_var = var_f("scan_var_" + str(cur_scan_idx), e.width)
+        f_res = e.f(cur_scan_lambda_var)
+        e_ret = sharing_nested_counters_rewrite(f_res, False, lake_state)
+        e = scan_const_f(lambda z: e_ret)
+
+    elif e_type in (AddOp, SubOp, ModOp, EqOp, LTOp, GTOp, MulOp):
+        e.arg0_index = get_index(e.arg0_index, lake_state)
+        e.arg1_index = get_index(e.arg1_index, lake_state)
+        lake_state.program_map[e.index] = e
+        merged = get_merged_counter(lake_state)
```

Comment:
please confirm my understanding: we're not modifying the addop/subop/mulop in this pass. We're only checking if their input is a merged counter, right? Because then we need to modify the input to these nodes?

Reply:
Yes, exactly, that's right.

Reply:
A note though to follow up on that: the merged counter val actually replaces the add op when you are adding two counter vals that are now merged, so that is replaced here.

Reply:
Makes sense. But why are you doing the same for mul and div? The rewrite rule only applies for add, right?

Reply:
That’s right, that’s a bug! Will update AddOp to be it’s own if statement with the merged code shortly.

Reply:
Great! Approving. Please merge this once you've fixed that.

Reply:
Pushed!